src/net_processing.cpp

```diff
@@ -187,6 +187,51 @@ struct QueuedBlock {
     std::unique_ptr<PartiallyDownloadedBlock> partialBlock;
 };
 
+class TimeOffsets
+{
+private:
+    // Historic maximum number of time offsets we include in our median.
+    static constexpr size_t N{199};
+
+    mutable Mutex m_mutex;
+    std::array<int64_t, N> m_offsets GUARDED_BY(m_mutex){};
+    size_t m_index GUARDED_BY(m_mutex){0};
+
+public:
+    /** Add a new time offset sample. */
+    void Add(int64_t offset) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        LOCK(m_mutex);
+
+        // Stop sampling after N samples.
+        // TODO This maintains a historic bug of adjusted time and could be
+        // changed in the future.
+        if (m_index >= N) return;
+
+        m_offsets[m_index] = offset;
+        ++m_index;
+    }
+
+    /** Compute and return the median of the collected time offset samples. */
+    int64_t Median() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        LOCK(m_mutex);
+
+        // Only compute the median from 5 or more samples.
+        if (m_index < 4) return 0;
+
+        std::array<int64_t, N> sorted_copy = m_offsets;
+        std::sort(sorted_copy.begin(), sorted_copy.begin() + m_index);
+
+        if (m_index % 2 == 0) {
+            return (sorted_copy[m_index / 2] / 2) +
```

Comment:
I think the the addition should be done before the division to handle rounding effects better, like it's done in `timedata`. If the two relevant offsets are, for example, 3 and 5, the median with this code would be 3.

Reply:
The current approach addresses overflow concerns as per https://github.com/bitcoin/bitcoin/pull/28956#discussion_r1411821098

Reply:
ok, I see. Still, having a function  that returns 3 as the medium of {3,5} and 5 as the medium of {4,6} seems inconsistent, couldn't we explicitly check and handle overflows instead?

Reply:
We already assume the median can be rounded down, so I don't think extra rounding changes any assumptions, and therefore I wouldn't overcomplicate things, it's just not worth it imo. I'd be okay with casting to a double first, but it's not strictly better as we're trading off precision at the higher ranges (which I also don't think is important at all).

```cpp
        if (m_index % 2 == 0) {
            return (0.5 * sorted_copy[m_index / 2]) +
                   (0.5 * sorted_copy[m_index / 2 - 1]);
        }
```

Reply:
Just a code comment would be fine too. I agree that the rounding doesn't matter at all in this context (and it'd also be fine not to calculate the mean in the even case at all), but code tends to get copied and reused.